lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -0,0 +1,199 @@
+# -*- mode: python -*-
+
+"""Tool Meister "Kill"
+
+Module responsible for hunting down and stopping all Tool Meisters, the Tool
+Data Sink, and the Redis server orchestrated by pbench-tool-meister-start.
+
+This is a "big hammer" approach that is offered to users when they find the
+distributed system state of the Pbench Agent not working correctly.
+
+The algorithm is fairly straight-forward:
+
+  For each pbench run directory in ${pbench_run}
+  1. If the run was NOT orchestrated by pbench-tool-meister-start, ignore
+  2. Find the recorded pids for the Redis Server, Tool Data Sink, and local
+     Tool Meister in their respective pid files, and stop those processes
+     from running
+  3. Determine all the remote hosts used for that run
+  4. For each remote host:
+     a. `ssh` to that remote host
+     b. Stop the Tool Meister running on that host
+
+The pbench-tool-meister-start uses a UUID in the command line of all the
+remote Tool Meister processes. Any Tool Meister process with that UUID in its
+command line string will be stopped via `kill -KILL`, along with all of its
+child processes.
+"""
+
+import click
+from collections import defaultdict
+import os
+import pathlib
+import psutil
+import shlex
+from typing import Tuple, List
+
+from pbench.agent.tool_group import ToolGroup
+from pbench.agent.utils import LocalRemoteHost, TemplateSsh
+from pbench.cli.agent import pass_cli_context, CliContext
+from pbench.cli.agent.base import BaseCommand
+from pbench.cli.agent.options import common_options
+
+
+class PidSource:
+    def __init__(self, file_name: str, display_name: str):
+        self.file_name = file_name
+        self.display_name = display_name
+        self.pids_by_uuid = {}
+        self.uuid_to_tmdir = {}
+
+    def load(self, tm_dir: pathlib.Path, uuid: str) -> bool:
+        """Load a PID from the given directory associated with the UUID given.
+
+        Records the loaded PID if it has a live process associated with it and
+        returns True, otherwise returns False.
+        """
+        try:
+            pid = (tm_dir / self.file_name).read_text()
+        except FileNotFoundError:
+            return False
+        try:
+            self.pids_by_uuid[uuid] = psutil.Process(pid)
+        except psutil.NoSuchProcess:
+            return False
+        self.uuid_to_tmdir[uuid] = tm_dir
+        return True
+
+    def killem(self):
+        """Kill all PIDs found, and their children."""
+        if not self.pids_by_uuid:
+            return
+        print(f"Killing {self.display_name} PIDs ...")
+        for uuid, proc in self.pids_by_uuid:
+            print(f"\tKilling {proc.pid} (from {self.uuid_to_tmdir[uuid]})")
+            # First get the list of children so that we can kill the parent
```

Comment:
Or maybe "... so we can then kill the parent first to prevent it from creating any other children."